src/common/Log.cpp

```diff
@@ -85,27 +85,94 @@ namespace Log {
         }
     }
 
-	//TODO add the time (broken for now because it is journaled) use Sys_Milliseconds instead (Utils::Milliseconds ?)
-    static const int warnTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE) | (1 << LOGFILE);
-    void CodeSourceWarn(std::string message) {
-        Log::Dispatch({"^3Warn: " + message}, warnTargets);
-    }
-
+    static const int debugTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE);
+    static const int verboseTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE) | (1 << LOGFILE);
     static const int noticeTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE) | (1 << LOGFILE);
-    void CodeSourceNotice(std::string message) {
-        Log::Dispatch({message}, noticeTargets);
-    }
+    static const int warnTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE) | (1 << LOGFILE);
 
-    static const int verboseTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE) | (1 << LOGFILE);
-    void CodeSourceVerbose(std::string message) {
-        Log::Dispatch({message}, verboseTargets);
+    //TODO add the time (broken for now because it is journaled) use Sys_Milliseconds instead (Utils::Milliseconds ?)
+    void DispatchByLevel(std::string message, Log::Level level) {
+        switch (level) {
+        case Level::DEBUG:
+            Log::Dispatch({"^5Debug: " + message}, debugTargets);
+            break;
+        case Level::VERBOSE:
+            Log::Dispatch({std::move(message)}, verboseTargets);
+            break;
+        case Level::NOTICE:
+            Log::Dispatch({std::move(message)}, noticeTargets);
+            break;
+        case Level::WARNING:
+            Log::Dispatch({"^3Warn: " + message}, warnTargets);
+        }
     }
 
-    static const int debugTargets = (1 << GRAPHICAL_CONSOLE) | (1 << TTY_CONSOLE);
-    void CodeSourceDebug(std::string message) {
-        Log::Dispatch({"^5Debug: " + message}, debugTargets);
+    namespace {
+        // Log-spam suppression: if more than MAX_OCCURRENCES log messages with the same format string
+        // are sent in less than INTERVAL_MS milliseconds, they will stop being printed.
+        // (Unless the spammy message desists long enough to be flushed out of the buffer: then it can
+        // be printed again.)
+        class LogSpamSuppressor {
```

Comment:
So for logs from VMs, There will be no suppression for messages sent within the same frame? (Admittedly, this is pretty useless though.

Reply:
Huh? Why wouldn't it work on logs within the same frame?

Reply:
nvm